### ISSUE TITLE
[QUESTION] Abot rsync processes opened by lsyncd

### ISSUE THREAD

**Issue by Episodio1** [axkibe/lsyncd#415]
Hi!

With this configuration:

```
settings {
        logfile = "/var/log/lsyncd/lsyncd.log",
        statusFile = "/var/log/lsyncd/lsyncd.status",
        inotifyMode = "CloseWrite",
        maxProcesses = 3,
        insist = true
}
sync{
        default.rsync,
        source="/home/admin/xxx/xxx",
        target="192.168.X.X:/home/admin/xxx/xxx",
        delay = 5,
        delete = false,
        exclude={ '*.wav' , '.*' },
        rsync  = {
                archive  = true,
                update   = true
        }
}
```

I am getting one rsync processes reading hard drive at maximum (5 MB/s) for a long time (There are millions of files in source folder). When that process  finishes (2-3 hours) lsyncd transfers the new files (around 3000 files which has been delayed), and then starts the same long-running process again.
When I run `ps aux |grep rsync` :
```
root      7190  6.3  0.0  12720   708 ?        D    06:17   0:25 /usr/bin/rsync -gsoputlD -r --force --from0 --include-from=- --exclude=* /home/admin/xxx/xxx/ 192.168.X.X:/home/admin/xxx/xxx/
root      7191  0.0  0.1  41444  1212 ?        S    06:17   0:00 ssh 192.168.X.X rsync --server -sulogDtpre.iLsf
root      7222  0.0  0.0  12300   424 ?        Ss   06:17   0:00 rsync --server -sulogDtpre.iLsf
```

Is the 1st process running a full scan of my folder?? 

Should I tweak "MaxProcesses" or "MaxDelays"?

Thanks.

**Comment by axkibe** [comment]
Look at the Lsyncd logs, if Lsyncd Queue overflowed during the initial rsync, yes it will run a full rsync again.

If not, albeit the second process looks the same in 'ps ax' it is in fact not, as Lsyncd passes filter rules via stdin pipe to rsync to tell it what files changed in the meantime, so it ought to run way faster than the inital rsync.

**Comment by Episodio1** [comment]
There are no errors in lsyncd.log. There are only:

```
Wed Dec 28 11:28:58 2016 Normal: Calling rsync with filter-list of new/modified files/dirs
file list...
Wed Dec 28 11:47:11 2016 Normal: Finished a list after exitcode: 0
Wed Dec 28 11:47:17 2016 Normal: Finished a list after exitcode: 0
Wed Dec 28 11:47:17 2016 Normal: Calling rsync with filter-list of new/modified files/dirs
file list...
Wed Dec 28 12:05:23 2016 Normal: Finished a list after exitcode: 0
Wed Dec 28 12:05:23 2016 Normal: Calling rsync with filter-list of new/modified files/dirs
file list...

```
I just notice there are a few times 2 consecutive `"Finished a list after exitcode: 0"`  from time to time. 


The destination node is running lsyncd as well, rsync'ing files to the source node on the same folder. I dont' see any sync loop problem.

**Comment by axkibe** [comment]
You got an external tmp dir and set rsync update true?

**Comment by Episodio1** [comment]
2 nodes:
- Node 1 with Virtual-IP receives the new files and lsyncd copies the files to node 2.
- Node 2 runs lsyncd and copy to 1st node, but files won't be transferred because they already exist in Node 1 (hidden files created bu rsync are excluded in my configuration, thuse avoiding loop and repeated copies between both nodes).
- When VirtualIP moves to Node 2, everything keeps running smoothly (both nodes have same files).
- There are no deletion in any nodes (delete = false).

"update" option from Rsync manual:

> This forces rsync to **skip any files which exist on the destination and have a modified time that is newer than the source file**. (If an existing destination file has a modification time equal to the source file's, it will be updated if the sizes are different.)

Do you think if I remove "update" it won't start new full scan again and again?

**Comment by axkibe** [comment]
No, you need update, otherwise you have loop, thats why I asked.

Having multiple "finished a list", message is normal, since you have increased maxProcesses aboth the default 1, so you have two rsyncs running in parallel.

For me everything looks as designed and working.

Otherwise try turning on "Delay" messaging to see why Lsyncd gives said file lists to rsync and there is something happening the way you didn't want to.

**Comment by Episodio1** [comment]
What if... I start lsyncd, it starts full scan (taking 2-3 hours), and meanwhile, there are new files being created (3000 files after those 2-3 hours)... perhaps Lsyncd doesn't behave as expected and it starts full scan again until there are no new files being created?

**Comment by axkibe** [comment]
Well look at the log files, it says with what file filters it (re)starts rsync.

It does a full resync when the inotify queue has overflowed and it missed events. But you should see that in the logs.

**Comment by axkibe** [comment]
PS: For debugging it might be simpler to set maxProcesses back down to 1.

**Comment by Episodio1** [comment]
Beofre restarting lsyncd service, If I change lsyncd config to add a new "_extra" parameter ("--stats" and "-h") to check rsync's output in every iteration:
```
 rsync  = {
                archive  = true,
                update   = true,
                _extra = {"--stats", "-h"}
        }
```
Do I have to stop/start lsyncd service to take effect?

**Comment by axkibe** [comment]
As you probably found out by now, yes.

**Comment by Episodio1** [comment]
Thanks axkibe.
Maybe this information helps to debug my situation...

After restarting lsyncd, it does a full scan:


```
Wed Jan  4 10:52:23 2017 Normal: recursive startup rsync: /home/admin/xxx/xxx/ -> 192.168.XXX.XXX:/home/admin/xxx/xxx/ excluding
*.wav
.*

Number of files: 2370796
Number of files transferred: 132
Total file size: 217.19G bytes
Total transferred file size: 21.84M bytes
Literal data: 21.84M bytes
Matched data: 0 bytes
File list size: 139.12M
File list generation time: 193.968 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 160.97M
Total bytes received: 2.54K

sent 160.97M bytes  received 2.54K bytes  111.28K bytes/sec
total size is 217.19G  speedup is 1349.23
Wed Jan  4 11:16:29 2017 Normal: Startup of /home/admin/xxx/xxx/ -> 192.168.XXX.XXX:/home/admin/xxx/xxx/ finished.
```

From 217 GB only 22 MB are transferred, almost all files are in destination, OK
 It took around 5 minutes real time.)

Then new files are created...

```
Wed Jan  4 11:25:03 2017 Normal: Calling rsync with filter-list of new/modified files/dirs
/2017/01/04/
/2017/01/
/2017/
/
/2017/01/04/2017-01-04_10-33-35-09265111-1033-1483536815561.ogg
/2017/01/04/2017-01-04_10-33-35-09265211-1033-1483536815531.ogg
/2017/01/04/2017-01-04_10-33-35-09265311-1033-1483536815584.ogg
...
...
/2017/01/04/2017.01.04.-10.58.47-2622689747.ogg
/2017/01/04/2017.01.04.-10.58.55-2615131161.ogg
```

LOG stops here showing nothing for a long time (5 minutes).
If I run "iotop" at this momento it shows high I/O reading usage for 1 rsync process.
Then resync spits output:

```
Number of files: 221
Number of files transferred: 0
Total file size: 20.25M bytes
Total transferred file size: 0 bytes
Literal data: 0 bytes
Matched data: 0 bytes
File list size: 10.68K
File list generation time: 280.033 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 10.70K
Total bytes received: 18

sent 10.70K bytes  received 18 bytes  38.07 bytes/sec
total size is 20.25M  speedup is 1889.18
Wed Jan  4 11:29:44 2017 Normal: Finished a list after exitcode: 0
```

It took 5 minutes to generate a 221-file list??

Next new files created...

```
Wed Jan  4 11:33:03 2017 Normal: Calling rsync with filter-list of new/modified files/dirs
/2017/01/04/
/2017/01/
/2017/
/
/2017/01/04/2017-01-04_11-07-54-09265111-1033-1483538874654.ogg
...
...
/2017/01/04/2017.01.04.-11.22.37-1159846300.ogg

Number of files: 48
Number of files transferred: 0
Total file size: 10.35M bytes
Total transferred file size: 0 bytes
Literal data: 0 bytes
Matched data: 0 bytes
File list size: 2.42K
File list generation time: 164.101 seconds
File list transfer time: 0.000 seconds
Total bytes sent: 2.44K
Total bytes received: 18

sent 2.44K bytes  received 18 bytes  14.83 bytes/sec
total size is 10.35M  speedup is 4218.45
Wed Jan  4 11:35:48 2017 Normal: Finished a list after exitcode: 0
```
Again 3-4 minutes for transferring a few MBytes.

**"file list generation time"** shows too much time for a 48 filename list, doesnt it ?


Any idea on what could be the problem?


Thanks you for your help.

**Comment by axkibe** [comment]
Can you redo this with '-log all' (as Lsyncd runtime parameter) and send me the complete logfile to axkibe@gmail.com please?

**Comment by Episodio1** [comment]
Thanks. Mail sent.

**Comment by axkibe** [comment]
Fixed in 2.2.1
